### PR TITLE
Technical Challenge 2 - Event Recommendation System

### PR DESCRIPTION
**Pull request by fiyinfoluwaafol**
### Description
This PR consists of the implementation of the feature that allows the user to be recommended time slots that to schedule events based on their availability, as well as the availability of invitees (applicable for shared events). 

The algorithm works by creating a general time grid, upon which slots will be marked as available or unavailable depending on whether the user or event invitees have events during that time slot. Once the time grid has been populated according to the events in the schedule of the user (as well invited users), it runs through the time grid and returns a list of available slots, while accounting for the desired duration of the event.

Although, I have made sure to test that this works in an ideal scenario, subsequent PRs can expect to be primarily about resolving bugs that come along the way during the final polishing and testing to ensure as many edge cases are accounted for.

Would also appreciate the feedback about other use cases that I haven't been considered with this current implementation.

### Milestones
This works towards the following user story:
User is given recommendations for event scheduling

### Test Plan
Here is the video showing this PR's test plan (including all discussed features):
[Technical Challenge 2 - Event Recommendation System](https://www.loom.com/share/5eb5b257d86641369df0d9c0cfce75b6?sid=5451dd03-1cc4-450b-8809-985591afa190)

### CODE REVIEW

File: backend/recommendation.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
+const { PrismaClient } = require('@prisma/client');
+const prisma = new PrismaClient();
+
+function generateTimeGrid() {
+    const timeGrid = [];
+    for (let i = 0; i < 48; i++) {
+        timeGrid.push({ slot: i, available: true });
```
Comment (jameskim08) [inline backend/recommendation.js:7]: What is the slot key for? It looks like you only really need the available key.

File: backend/recommendation.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    return timeGrid;
+}
+
+function markUnavailableSlots(timeGrid, startAt, endAt) {
+    const startSlot = Math.floor(startAt.getHours() * 2 + startAt.getMinutes() / 30);
```
Comment (jameskim08) [inline backend/recommendation.js:13]: a comment here may help understand this part. From what I understand, each hour takes up two slots in your time grid, and every 30 minutes take up 1 slot?

File: frontend/src/components/CreateEvent/CreateEvent.jsx [hunk 3]
```diff
@@ -69,6 +93,27 @@ function CreateEvent ({onClose, onCreate, onEdit, initialView, isEdit}) {
                     <p>Date</p>
                     <input type="date" placeholder="mm/dd/yyyy" value={date} onChange={(e) => setDate(e.target.value)} required/>
 
+                    <p>Duration</p>
+                    <input type="number" value={duration} onChange={(e) => setDuration(e.target.value)} placeholder="Duration in 30-min slots" min="1" />
```
Comment (jameskim08) [inline frontend/src/components/CreateEvent/CreateEvent.jsx:97]: From a user pov, the duration is pretty unclear. You should provide options of 30 minute increments.

File: frontend/src/components/CreateEvent/CreateEvent.jsx [hunk 4]
```diff
@@ -69,6 +93,27 @@ function CreateEvent ({onClose, onCreate, onEdit, initialView, isEdit}) {
...
+                    <button type="button" onClick={fetchRecommendations}>Get Recommendations</button>
+                    {recommendations.length > 0 && (
+                        <div>
+                            <h2>Recommended Times</h2>
+                            <ul>
+                            {recommendations.map((rec, index) => {
+                                const startTime = slotToTime(rec);
+                                const endTime = slotToTime(rec + duration);
+                                return (
+                                    <li key={index} onClick={() => {
```
Comment (jameskim08) [inline frontend/src/components/CreateEvent/CreateEvent.jsx:107]: nice, you made these clickable! I would suggest making these more obviously buttons, as it's not very clear they are right now.

File: backend/recommendation.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        const users = await prisma.user.findMany({
+            where: {
+                OR: [{ id: userId }, { email: { in: invitees } }]
+            },
+            include: {
+                calendarEvents: {
+                    where: {
+                        status: 'accepted',
+                        AND: [
+                            { startAt: { lte: endOfDay } },
```
Comment (jameskim08) [inline backend/recommendation.js:51]: sorry, this might be a noob question but why is endOfDay set here for startAt? Same with the line below in reverse

File: backend/recommendation.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            }
+        });
+
+        if (users.length === 0) {
+            throw new Error('No users found');
+        }
+
+        let timeGrid = generateTimeGrid();
+        users.forEach(user => {
+            const userTimeGrid = generateTimeGrid();
```
Comment (jameskim08) [inline backend/recommendation.js:65]: Why does each user need their own separate time grid? Can we just update the general time grid as we go through each user?

File: backend/recommendation.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            }
+            for (let i = user.preferredEndTime; i < userTimeGrid.length; i++) {
+                userTimeGrid[i].available = false;
+            }
+            timeGrid = timeGrid.map((slot, index) => ({ slot: index, available: slot.available && userTimeGrid[index].available }));
+        });
+
+        const availableSlots = findAvailableSlots(timeGrid, duration);
+        if (availableSlots.length === 0) {
+            throw new Error('No available slots found');
```
Comment (jameskim08) [inline backend/recommendation.js:82]: This shouldn't be an error, just return an empty array and let the UI handle it (some message saying there's no available time slots).

File: frontend/src/components/CreateEvent/CreateEvent.jsx [hunk 8]
```diff
@@ -10,7 +11,30 @@ function CreateEvent ({onClose, onCreate, onEdit, initialView, isEdit}) {
     const [location, setLocation] = useState('');
     const [allDay, setAllDay] = useState(false);
     const [participants, setParticipants] = useState('');
-
+    const [duration, setDuration] = useState(1);
+    const [recommendations, setRecommendations] = useState([]);
+
+    const fetchRecommendations = async () => {
```
Comment (jameskim08) [inline frontend/src/components/CreateEvent/CreateEvent.jsx:17]: you should move the fetching portion of this function to utils, so it's not recreated on every render

File: backend/recommendation.js (hunk 5 above)
Comment (fiyinfoluwaafol) [reply backend/recommendation.js:51]: Apologies for not leaving descriptive enough comments here, what I am doing here is trying to get the calendar events which have a startAt that is less than endOfDay (which means that start before 11:59PM of the specified day) and also have an endAt that is greater than startOfDay (which means that end after 12:00AM of the specified day). When these two conditions are combined, it gives a list of calendar events that are essentially within the range of the specified day.

Hopefully that clarifies what I did
